src/main/kotlin/org/elm/workspace/ElmFormatUtils.kt

```diff
@@ -0,0 +1,69 @@
+package org.elm.workspace
+
+import com.intellij.execution.ExecutionException
+import com.intellij.execution.process.ProcessOutput
+import com.intellij.openapi.application.ApplicationManager
+import com.intellij.openapi.command.CommandProcessor
+import com.intellij.openapi.editor.Document
+import com.intellij.openapi.fileEditor.FileDocumentManager
+import com.intellij.openapi.progress.ProgressManager
+import com.intellij.openapi.project.Project
+import com.intellij.openapi.project.ProjectManager
+import com.intellij.openapi.roots.ProjectRootManager
+import com.intellij.openapi.vfs.VirtualFile
+import com.intellij.psi.PsiFile
+import org.elm.lang.core.psi.ElmFile
+import org.elm.openapiext.isSuccess
+import org.elm.openapiext.toPsiFile
+
+object ElmFormatUtils {
+
+    fun getElmVersion(project: Project, file: VirtualFile): Version? {
+        val psiFile = file.toPsiFile(project) ?: return null
+        return getElmVersion(psiFile)
+    }
+
+    fun getElmVersion(psiFile: PsiFile): Version? {
+        if (psiFile !is ElmFile) return null
+        val elmVersion: Version = when (val elmProject = psiFile.elmProject) {
+            is ElmApplicationProject -> elmProject.elmVersion
+            is ElmPackageProject -> elmProject.elmVersion.low
+            else -> return null
+        }
+
+        return elmVersion
+    }
+
+    fun formatDocumentAndSetText(project: Project, document: Document, elmFormatCLI: ElmFormatCLI) {
+
+        val file = FileDocumentManager.getInstance().getFile(document)
+        val version = getElmVersion(project, file!!)
+        val result = ProgressManager.getInstance().runProcessWithProgressSynchronously<ProcessOutput, ExecutionException>({
+            elmFormatCLI.reformatFile(project, version!!, document)
+        }, "Running elm-format on current file...", true, project)
+
+        if (result.isSuccess) {
+            val formatted = result.stdout
+            val source = document.text
+
+            if (source != formatted) {
+                ApplicationManager.getApplication().runWriteAction {
+                    CommandProcessor
+                            .getInstance()
+                            .runUndoTransparentAction {
+                                document.setText(formatted)
+                            }
+                }
+            }
+        }
+    }
+
+    fun projectForFile(file: VirtualFile): Project? {
+        return ProjectManager.getInstance().openProjects.find { project ->
+            val fileIndex = ProjectRootManager.getInstance(project).fileIndex
+            // FIXME source-directories listed in elm.json aren't marked as Source Root in the IDEA project
```

Comment:
Yeah, the 'source root' thing was never implemented. But just today I started working on some stuff related to that (#262). It doesn't actually do the work of marking src/test/excluded roots, but that will be coming later.

Reply:
nice!